src/powershell/Program.cs

```diff
@@ -19,7 +63,458 @@ public sealed class ManagedPSEntry
+                IntPtr bufSize = ReadLink("/proc/self/exe", linkPathPtr, (UIntPtr)LINUX_PATH_MAX);
+                pwshPath = Marshal.PtrToStringAnsi(linkPathPtr, (int)bufSize);
+                Marshal.FreeHGlobal(linkPathPtr);
+
+                // exec pwsh
+                ThrowOnFailure("exec", ExecPwshLogin(args, pwshPath, isMacOS: false));
+                return;
+            }
+
+            // At this point, we are on macOS
+
+            // Set up the mib array and the query for process maximum args size
+            Span<int> mib = stackalloc int[3];
+            int mibLength = 2;
+            mib[0] = MACOS_CTL_KERN;
+            mib[1] = MACOS_KERN_ARGMAX;
+            int size = IntPtr.Size / 2;
+            int argmax = 0;
+
+            // Get the process args size
+            unsafe
+            {
+                fixed (int *mibptr = mib)
+                {
+                    ThrowOnFailure(nameof(argmax), SysCtl(mibptr, mibLength, &argmax, &size, IntPtr.Zero, 0));
+                }
+            }
+
+            // Get the PID so we can query this process' args
+            int pid = GetPid();
+
+            // The following logic is based on https://gist.github.com/nonowarn/770696
+
+            // Now read the process args into the allocated space
+            IntPtr procargs = Marshal.AllocHGlobal(argmax);
+            IntPtr execPathPtr = IntPtr.Zero;
+            try
+            {
+                mib[0] = MACOS_CTL_KERN;
+                mib[1] = MACOS_KERN_PROCARGS2;
+                mib[2] = pid;
+                mibLength = 3;
+
+                unsafe
+                {
+                    fixed (int *mibptr = mib)
+                    {
+                        ThrowOnFailure(nameof(procargs), SysCtl(mibptr, mibLength, procargs.ToPointer(), &argmax, IntPtr.Zero, 0));
+                    }
+
+                    // The memory block we're reading is a series of null-terminated strings
+                    // that looks something like this:
+                    //
+                    // | argc      | <int>
+                    // | exec_path | ... \0
+                    // | argv[0]   | ... \0
+                    // | argv[1]   | ... \0
+                    //   ...
+                    //
+                    // We care about argv[0], since that's the name the process was started with.
+                    // If argv[0][0] == '-', we have been invoked as login.
+                    // Doing this, the buffer we populated also recorded `exec_path`,
+                    // which is the path to our executable `pwsh`.
+                    // We can reuse this value later to prevent needing to call a .NET API
+                    // to generate our exec invocation.
+
+
+                    // We don't care about argc's value, since argv[0] must always exist.
+                    // Skip over argc, but remember where exec_path is for later
+                    execPathPtr = IntPtr.Add(procargs, sizeof(int));
```

Comment:
The variable name `execPathPtr` is misleading as we are calling `exec` below. The var name seems indicating it holds the path to the `exec` utility.
I suggest to change it to `executablePathPtr`.

Reply:
Well it's actually a pointer to `exec_path`; I was hoping my use of `Path` would indicate that. But I can change the name

Reply:
Yes, I understand it's pointing to `exec_path`, and it's natural to name the var as `execPathxxx` with no question. The thing is, in our situation here, we are invoking `exec` subsequently, and `exec_path` makes it look like _the path to exec_ whereas it actually means _the path of the executable_ and that's the part I found confusing when looking at the code initially.

Reply:
Well `exec` doesn't have a path; it's either a builtin or a syscall. It lives only in memory

Reply:
True, maybe it's just because I'm not familiar with whether `exec` is an utility or a builtin shell command :)